### PR TITLE
DPE-2124 - refactor - in lib config file render

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

* testing profile and sub 2GB ram missing additional memory savings (DPE-2124)
* mysqld configuration file is (now) identically rendered for k8s and vm, but implementation is separated

## Solution

* added extra memory savings
* mysqld configuration file render in library (memory calc and write to system on each charm)
* couple of type fixes
* decreased dns observer log level (info->debug)

### CODE REVIEW

**Review by taurus-forever — DISMISSED** [review]
Great step forward!

File: lib/charms/mysql/v0/mysql.py [hunk 1]
```diff
@@ -375,6 +379,11 @@ def __init__(self, *args):
         # of methods
         self.current_event = None
 
+    @abstractproperty
+    def _mysql(self) -> "MySQLBase":
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py:383]: MySQLBase relates to the workload while MySQLCharmBase to the charm.

File: lib/charms/mysql/v0/mysql.py [hunk 2]
```diff
@@ -461,15 +470,15 @@ def unit_initialized(self):
         return self.unit_peer_data.get("unit-initialized") == "True"
 
     @property
-    def app_peer_data(self) -> Dict:
+    def app_peer_data(self) -> ops.RelationDataContent:
         """Application peer relation data object."""
         if self.peers is None:
             return {}
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py]: 6b9229c

File: lib/charms/mysql/v0/mysql.py [hunk 3]
```diff
@@ -375,6 +379,11 @@ def __init__(self, *args):
         # of methods
         self.current_event = None
 
+    @abstractproperty
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py]: 6b9229c

File: lib/charms/mysql/v0/mysql.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py:384]: 6b9229c

File: lib/charms/mysql/v0/mysql.py [hunk 4]
```diff
@@ -1959,9 +2020,14 @@ def offline_mode_and_hidden_instance_exists(self) -> bool:
...
         return matches.group(1) != "0"
 
-    def get_innodb_buffer_pool_parameters(self) -> Tuple[int, Optional[int], Optional[int]]:
+    def get_innodb_buffer_pool_parameters(
+        self, available_memory: int
+    ) -> Tuple[int, Optional[int], Optional[int]]:
         """Get innodb buffer pool parameters for the instance.
 
+        Args:
+            available_memory: The amount of memory available to the instance
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py]: 6b9229c

File: src/mysql_vm_helpers.py [hunk 5]
```diff
@@ -183,56 +183,47 @@ def install_and_configure_mysql_dependencies() -> None:
...
+        try:
+            content = self.render_myqld_configuration(profile=profile)
+
+            # create the mysqld config directory if it does not exist
+            pathlib.Path(MYSQLD_CONFIG_DIRECTORY).mkdir(mode=0o755, parents=True, exist_ok=True)
 
-        with open(f"{MYSQLD_CONFIG_DIRECTORY}/z-custom-mysqld.cnf", "w") as config_file:
-            config_file.write("\n".join(content))
+            with open(f"{MYSQLD_CONFIG_DIRECTORY}/z-custom-mysqld.cnf", "w") as config_file:
+                config_file.write(content)
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/mysql_vm_helpers.py]: 6b9229c

File: src/mysql_vm_helpers.py [hunk 6]
```diff
@@ -7,25 +7,25 @@
...
     MySQLGetAutoTunningParametersError,
+    MySQLGetAvailableMemoryError,
     MySQLRestoreBackupError,
     MySQLServiceNotRunningError,
     MySQLStartMySQLDError,
     MySQLStopMySQLDError,
 )
 from charms.operator_libs_linux.v1 import snap
 from tenacity import retry, stop_after_delay, wait_fixed
+from typing_extensions import override
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/mysql_vm_helpers.py:28]: 6b9229c

File: lib/charms/mysql/v0/mysql.py [hunk 7]
```diff
@@ -111,7 +111,7 @@ def wait_until_mysql_connection(self) -> None:
 
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
-LIBPATCH = 42
+LIBPATCH = 43
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/mysql.py:114]: note to confirm actual libpatch (it is greater than 43 now i believe)

**Review by shayancanonical — DISMISSED** [review]

**Review by shayancanonical — APPROVED** [review]

**Review by taurus-forever — DISMISSED** [review]
Nice direction!
re, info->debug: https://warthogs.atlassian.net/browse/DPE-2455 :-D

File: poetry.lock [hunk 8]
```diff
@@ -13,17 +13,17 @@ files = [
 
 [[package]]
 name = "asttokens"
-version = "2.2.1"
+version = "2.4.0"
```
[the remark this answers is not in the document]
Comment (paulomach) [reply poetry.lock:16]: probably at some point - good call

[the remark this answers is not in the document]
Comment (paulomach) [reply poetry.lock]: 5234117

**Review by shayancanonical — APPROVED** [review]

**Pull request merged**